grep/grep.1
.Dd December 2008
.Dt GREP 1

.Sh NAME

.Nm grep
.Nd search a file for a pattern

.Sh SYNOPSIS

    grep [-E| -F][-c| -l| -q][-insvx] -e pattern_list...
           [-f pattern_file]...[file...]

    grep [-E| -F][-c| -l| -q][-insvx][-e pattern_list]...
           -f pattern_file...[file...]

    grep [-E| -F][-c| -l| -q][-insvx] pattern_list[file...]

.Sh DESCRIPTION

    The grep utility shall search the input files, selecting lines matching
one or more patterns; the types of patterns are controlled by the options
specified. The patterns are specified by the -e option, -f option, or the
pattern_list operand. The pattern_list's value shall consist of one or more
patterns separated by <newline>s; the pattern_file's contents shall consist
of one or more patterns terminated by <newline>. By default, an input line
shall be selected if any pattern, treated as an entire basic regular
expression (BRE) as described in the Base Definitions volume of IEEE Std
1003.1-2001, Section 9.3, Basic Regular Expressions, matches any part of the
line excluding the terminating <newline>; a null BRE shall match every line.
By default, each selected input line shall be written to the standard output.

    Regular expression matching shall be based on text lines. Since a
<newline> separates or terminates patterns (see the -e and -f options below),
regular expressions cannot contain a <newline>. Similarly, since patterns are
matched against individual lines (excluding the terminating <newline>s) of
the input, there is no way for a pattern to match a <newline> found in the
input.

.Sh OPTIONS

    The grep utility shall conform to the Base Definitions volume of IEEE Std
1003.1-2001, Section 12.2, Utility Syntax Guidelines.

    The following options shall be supported:

    -E
        Match using extended regular expressions. Treat each pattern
specified as an ERE, as described in the Base Definitions volume of IEEE Std
1003.1-2001, Section 9.4, Extended Regular Expressions. If any entire ERE
pattern matches some part of an input line excluding the terminating
<newline>, the line shall be matched. A null ERE shall match every line.
    -F
        Match using fixed strings. Treat each pattern specified as a string
instead of a regular expression. If an input line contains any of the
patterns as a contiguous sequence of bytes, the line shall be matched. A null
string shall match every line.
    -c
        Write only a count of selected lines to standard output.
    -e  pattern_list
        Specify one or more patterns to be used during the search for input.
The application shall ensure that patterns in pattern_list are separated by a
<newline>. A null pattern can be specified by two adjacent <newline>s in
pattern_list. Unless the -E or -F option is also specified, each pattern
shall be treated as a BRE, as described in the Base Definitions volume of
IEEE Std 1003.1-2001, Section 9.3, Basic Regular Expressions. Multiple -e and
-f options shall be accepted by the grep utility. All of the specified
patterns shall be used when matching lines, but the order of evaluation is
unspecified.
    -f  pattern_file
        Read one or more patterns from the file named by the pathname
pattern_file. Patterns in pattern_file shall be terminated by a <newline>. A
null pattern can be specified by an empty line in pattern_file. Unless the -E
or -F option is also specified, each pattern shall be treated as a BRE, as
described in the Base Definitions volume of IEEE Std 1003.1-2001, Section
9.3, Basic Regular Expressions.
    -i
        Perform pattern matching in searches without regard to case; see the
Base Definitions volume of IEEE Std 1003.1-2001, Section 9.2, Regular
Expression General Requirements.
    -l
        (The letter ell.) Write only the names of files containing selected
lines to standard output. Pathnames shall be written once per file searched.
If the standard input is searched, a pathname of "(standard input)" shall be
written, in the POSIX locale. In other locales, "standard input" may be
replaced by something more appropriate in those locales.
    -n
        Precede each output line by its relative line number in the file,
each file starting at line 1. The line number counter shall be reset for each
file processed.
    -q
        Quiet. Nothing shall be written to the standard output, regardless of
matching lines. Exit with zero status if an input line is selected.
    -s
        Suppress the error messages ordinarily written for nonexistent or
unreadable files. Other error messages shall not be suppressed.
    -v
        Select lines not matching any of the specified patterns. If the -v
option is not specified, selected lines shall be those that match any of the
specified patterns.
    -x
        Consider only input lines that use all characters in the line
excluding the terminating <newline> to match an entire fixed string or
regular expression to be matching lines.

.Sh OPERANDS

    The following operands shall be supported:

    pattern_list
        Specify one or more patterns to be used during the search for input.
This operand shall be treated as if it were specified as -e pattern_list.
    file
        A pathname of a file to be searched for the patterns. If no file
operands are specified, the standard input shall be used.

.Sh STDIN

    The standard input shall be used only if no file operands are specified.
See the INPUT FILES section.

.Sh INPUT FILES

    The input files shall be text files.

.Sh ENVIRONMENT VARIABLES

    The following environment variables shall affect the execution of grep:

    LANG
        Provide a default value for the internationalization variables that
are unset or null. (See the Base Definitions volume of IEEE Std 1003.1-2001,
Section 8.2, Internationalization Variables for the precedence of
internationalization variables used to determine the values of locale
categories.)
    LC_ALL
        If set to a non-empty string value, override the values of all the
other internationalization variables.
    LC_COLLATE
        Determine the locale for the behavior of ranges, equivalence classes,
and multi-character collating elements within regular expressions.
    LC_CTYPE
        Determine the locale for the interpretation of sequences of bytes of
text data as characters (for example, single-byte as opposed to multi-byte
characters in arguments and input files) and the behavior of character
classes within regular expressions.
    LC_MESSAGES
        Determine the locale that should be used to affect the format and
contents of diagnostic messages written to standard error.
    NLSPATH
        [XSI] [Option Start] Determine the location of message catalogs for
the processing of LC_MESSAGES . [Option End]

.Sh ASYNCHRONOUS EVENTS

    Default.

.Sh STDOUT

    If the -l option is in effect, and the -q option is not, the following
shall be written for each file containing at least one selected input line:

    "%s\n", <file>

    Otherwise, if more than one file argument appears, and -q is not
specified, the grep utility shall prefix each output line by:

    "%s:", <file>

    The remainder of each output line shall depend on the other options
specified:

        If the -c option is in effect, the remainder of each output line
shall contain:

        "%d\n", <count>

        Otherwise, if -c is not in effect and the -n option is in effect, the
following shall be written to standard output:

        "%d:", <line number>

        Finally, the following shall be written to standard output:

        "%s", <selected-line contents>

.Sh STDERR

    The standard error shall be used only for diagnostic messages.

.Sh OUTPUT FILES

    None.

.Sh EXTENDED DESCRIPTION

    None.

.Sh EXIT STATUS

    The following exit values shall be returned:

     0
        One or more lines were selected.
     1
        No lines were selected.
    >1
        An error occurred.

.Sh CONSEQUENCES OF ERRORS

    If the -q option is specified, the exit status shall be zero if an input
line is selected, even if an error was detected. Otherwise, default actions
shall be performed.

The following sections are informative.
.Sh APPLICATION USAGE

    Care should be taken when using characters in pattern_list that may also
be meaningful to the command interpreter. It is safest to enclose the entire
pattern_list argument in single quotes:

    '...'

    The -e pattern_list option has the same effect as the pattern_list
operand, but is useful when pattern_list begins with the hyphen delimiter. It
is also useful when it is more convenient to provide multiple patterns as
separate arguments.

    Multiple -e and -f options are accepted and grep uses all of the patterns
it is given while matching input text lines. (Note that the order of
evaluation is not specified. If an implementation finds a null string as a
pattern, it is allowed to use that pattern first, matching every line, and
effectively ignore any other patterns.)

    The -q option provides a means of easily determining whether or not a
pattern (or string) exists in a group of files. When searching several files,
it provides a performance improvement (because it can quit as soon as it
finds the first match) and requires less care by the user in choosing the set
of files to supply as arguments (because it exits zero if it finds a match
even if grep detected an access or read error on earlier file operands).

.Sh EXAMPLES

        To find all uses of the word "Posix" (in any case) in file text.mm
and write with line numbers:

        grep -i -n posix text.mm

        To find all empty lines in the standard input:

        grep ^$

        or:

        grep -v .

        Both of the following commands print all lines containing strings
"abc" or "def" or both:

        grep -E 'abc|def'


        grep -F 'abc
        def'

        Both of the following commands print all lines matching exactly "abc"
or "def" :

        grep -E '^abc$|^def$'


        grep -F -x 'abc
        def'

.Sh RATIONALE

    This grep has been enhanced in an upwards-compatible way to provide the
exact functionality of the historical egrep and fgrep commands as well. It
was the clear intention of the standard developers to consolidate the three
greps into a single command.

    The old egrep and fgrep commands are likely to be supported for many
years to come as implementation extensions, allowing historical applications
to operate unmodified.

    Historical implementations usually silently ignored all but one of
multiply-specified -e and -f options, but were not consistent as to which
specification was actually used.

    The -b option was omitted from the OPTIONS section because block numbers
are implementation-defined.

    The System V restriction on using - to mean standard input was omitted.

    A definition of action taken when given a null BRE or ERE is specified.
This is an error condition in some historical implementations.

    The -l option previously indicated that its use was undefined when no
files were explicitly named. This behavior was historical and placed an
unnecessary restriction on future implementations. It has been removed.

    The historical BSD grep -s option practice is easily duplicated by
redirecting standard output to /dev/null. The -s option required here is from
System V.

    The -x option, historically available only with fgrep, is available here
for all of the non-obsolescent versions.

.Sh FUTURE DIRECTIONS

    None.

.Sh SEE ALSO

    sed



NEWS

Version 5 - unreleased

	- New utilities: cmp, tput, tabs, stty, uname, uniq

	- Bug fixes.

Version 4 - December 9, 2006

	- import codebase into git repository on git.kernel.org
	- fix compiler warnings on newer compilers
	- add x86-64 support
	- add WIP utility: pax
	- add utilities: env, id, strings, time
	- add simple "it works" tests
	- create per-util status files
	- drop hand-rolled 'compress', import BSD 'compress' instead

Version 3 - February 7, 2005

	- Add copyright/license info to each source file.
	- Make sure bug report address points to SourceForge.

	- New utilities: grep, pathchk, pwd, split, tty.
	- New work-in-progress (WIP) utility: mv.

	- Fix many 'compress' bugs.
	- Security fixes.

	- (internal) Move compress engine into lib.

Version 2 - January 26, 2005

	- Update README with project goals/manifesto,
	  and some URLs.

	- New utilities: compress, comm, date, what, wc, renice,
	  ipcrm, kill, rm.

	- Bug fixes.

Version 1 - August 21, 2004

	- Initial release.



grep.c
/*
 * Copyright 2004-2006 Jeff Garzik <jgarzik@pobox.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#ifndef HAVE_CONFIG_H
#error missing autoconf-generated config.h.
#endif
#include "posixutils-config.h"

#ifndef _GNU_SOURCE
#define _GNU_SOURCE		/* for strcasestr(3), fgets_unlocked(3) */
#endif

#define __STDC_FORMAT_MACROS 1

#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <inttypes.h>
#include <limits.h>
#include <regex.h>
#include <libpu.h>


static const char doc[] =
N_("grep - print lines matching a pattern");

static const char args_doc[] = N_("[pattern_list] [file...]");

static struct argp_option options[] = {
	{ "extended-regexp", 'E', NULL, 0,
	  N_("Match using extended regular expressions") },
	{ "fixed-strings", 'F', NULL, 0,
	  N_("Match using fixed strings") },
	{ "count", 'c', NULL, 0,
	  N_("Write only a count of selected lines to standard output") },
	{ "regexp", 'e', "pattern_list", 0,
	  N_("Specify one or more patterns to be used during the search for input") },
	{ "file", 'f', "pattern_file", 0,
	  N_("Read one or more patterns from the file named by the pathname pattern_file") },
	{ "ignore-case", 'i', NULL, 0,
	  N_("Perform pattern matching in searches without regard to case") },
	{ "files-with-matches", 'l', NULL, 0,
	  N_("Write only the names of files containing selected lines to standard output") },
	{ "line-number", 'n', NULL, 0,
	  N_("Precede each output line by its relative line number in the file") },
	{ "quiet", 'q', NULL, 0,
	  N_("Quiet. Nothing shall be written to the standard output, regardless of matching lines. Exit with zero status if an input line is selected.") },
	{ "silent", 0, NULL, OPTION_ALIAS },
	{ "no-messages", 's', NULL, 0,
	  N_("Suppress the error messages ordinarily written for nonexistent or unreadable files") },
	{ "invert-match", 'v', NULL, 0,
	  N_("Select lines not matching any of the specified patterns") },
	{ "line-regexp", 'x', NULL, 0,
	  N_("Consider only input lines that use all characters in the line, excluding the terminating <newline>, to match an entire fixed string or regular expression to be matching lines") },
	{ }
};

static int grep_init(struct walker *w, int argc, char **argv);
static int grep_pre_walk(struct walker *w);
static int grep_post_walk(struct walker *w);
static int grep_file(struct walker *w, const char *pr_fn, FILE *f);
static error_t parse_opt (int key, char *arg, struct argp_state *state);
static const struct argp argp = { options, parse_opt, args_doc, doc };

static struct walker walker;

enum match_type {
	MATCH_BRE,		/* match basic regex */
	MATCH_ERE,		/* match extended regex */
	MATCH_STRING,		/* match bytestring rather than regex */
};

enum output_type {
	OUT_NONE,		/* no output */
	OUT_CONTENT,		/* output [non-]matching lines */
	OUT_PATH,		/* output a pathname */
	OUT_COUNT,		/* output a count of [non-]matching lines */
};

enum output_options {
	OUT_LINENO		= (1 << 0),	/* number output lines */
	OUT_SUPPRESS_ERRMSG	= (1 << 1),	/* suppress err msgs */
};

enum {
	INIT_PATTERNS		= 16,		/* pattern array start size */

	STOP_LOOP		= (1 << 30),
};

struct pattern {
	regex_t			rx;
	const char		*pat_str;
};


/* command line options */
static int opt_match = MATCH_BRE;
static int opt_output = OUT_CONTENT;
static int output_opts;
static bool opt_ignore_case;
static bool opt_match_invert;
static bool opt_match_line;

/* pattern array */
static struct pattern *patterns;
static int n_patterns;
static int alloc_patterns;

/* input buffer */
static char linebuf[LINE_MAX + 1];

/* various per-file or per-invocation counters */
static uintmax_t n_matches;
static uintmax_t total_matches;
static uintmax_t n_lines;
static int n_files;


static void compile_patterns(void)
{
	struct pattern *pat;
	int rc, i, flags;

	flags = REG_NOSUB;
	if (opt_ignore_case)
		flags |= REG_ICASE;
	if (opt_match == MATCH_ERE)
		flags |= REG_EXTENDED;

	for (i = 0; i < n_patterns; i++) {
		const char *pat_str;

		pat = &patterns[i];
		pat_str = pat->pat_str;

		/* if match-full-line, add ^ and $ to string */
		if (opt_match_line) {
			int need_pre, need_suff;
			size_t len = strlen(pat_str);

			need_pre = (pat_str[0] != '^');
			if (!len)
				need_suff = 1;
			else
				need_suff = (pat_str[len - 1] != '$');

			/* replace pattern with new string */
			if (need_pre || need_suff) {
				char *s = (char *) xmalloc(len + 3);
				sprintf(s, "%s%s%s",
					need_pre ? "^" : "",
					pat_str,
					need_suff ? "$" : "");
				pat_str = pat->pat_str = s;
			}
		}

		/* compile pattern */
		rc = regcomp(&pat->rx, pat_str, flags);
		if (rc) {
			regerror(rc, &pat->rx, linebuf, sizeof(linebuf));
			fprintf(stderr, _("invalid pattern '%s': %s\n"),
				pat_str, linebuf);
			exit(2);
		}
	}
}

static void add_pattern(const char *pat_str)
{
	struct pattern *pat;

	/* grow pattern array if necessary */
	if (alloc_patterns == n_patterns) {
		if (alloc_patterns == 0)
			alloc_patterns = INIT_PATTERNS;
		else
			alloc_patterns <<= 1;

		patterns = (struct pattern *) xrealloc(patterns,
				    alloc_patterns * sizeof(struct pattern));
	}

	/* add pattern to array */
	pat = &patterns[n_patterns];
	pat->pat_str = pat_str;

	n_patterns++;
}

static void add_patterns(const char *pattern_list_in)
{
	char *pattern_list = xstrdup(pattern_list_in);
	size_t len = strlen(pattern_list_in);

	while (len > 0) {
		char *nl;
		int i;

		/* search for newline */
		nl = (char *) memchr(pattern_list, '\n', len);

		/* if no newline, this is last pattern in list */
		if (!nl) {
			add_pattern(pattern_list);
			break;
		}

		/* null-terminate string at newline, add as pattern */
		*nl = 0;
		i = nl - pattern_list;
		add_pattern(pattern_list);

		/* advance to next pattern, if any */
		i++;	/* include newline now */
		pattern_list += i;
		len -= i;
	}
}

static void add_pattern_file(const char *fn)
{
	FILE *f;
	char *s;
	size_t len;

	if (ro_file_open(&f, fn))
		goto err_out_exit;

	setlinebuf(f);

	while ((s = fgets_unlocked(linebuf, sizeof(linebuf), f)) != NULL) {
		len = strlen(s);
		if (s[len - 1] == '\n') {
			len--;
			s[len] = 0;
		}

		add_pattern(xstrdup(s));
	}

	if (ferror(f))
		perror(fn);
	if (fclose(f))
		goto err_out;

	return;

err_out:
	perror(fn);
err_out_exit:
	exit(2);
}

static error_t parse_opt (int key, char *arg, struct argp_state *state)
{
	PU_OPT_BEGIN
	PU_OPT_SET('i', ignore_case)
	PU_OPT_SET('v', match_invert)
	PU_OPT_SET('x', match_line)

	case 'E':
		opt_match = MATCH_ERE;
		break;
	case 'F':
		opt_match = MATCH_STRING;
		break;
	case 'c':
		opt_output = OUT_COUNT;
		break;
	case 'l':
		opt_output = OUT_PATH;
		break;
	case 'n':
		output_opts |= OUT_LINENO;
		break;
	case 'q':
		opt_output = OUT_NONE;
		break;
	case 's':
		output_opts |= OUT_SUPPRESS_ERRMSG;
		break;

	case 'e':
		add_patterns(arg);
		break;
	case 'f':
		add_pattern_file(arg);
		break;

	PU_OPT_ARG
	PU_OPT_DEFAULT
	PU_OPT_END
}

static int match_string(const char *line)
{
	int i;

	/* attempt to match entire line */
	if (opt_match_line)
		for (i = 0; i < n_patterns; i++) {
			int rc;
			if (opt_ignore_case)
				rc = strcasecmp(line, patterns[i].pat_str);
			else
				rc = strcmp(line, patterns[i].pat_str);
			if (rc == 0)
				return 1;
		}

	/* attach to match substring within line */
	else
		for (i = 0; i < n_patterns; i++) {
			const char *s;
			if (opt_ignore_case)
				 s = strcasestr(line, patterns[i].pat_str);
			else
				 s = strstr(line, patterns[i].pat_str);
			if (s)
				return 1;
		}

	return 0;
}

static int match_regex(const char *line)
{
	int i;

	for (i = 0; i < n_patterns; i++)
		if (regexec(&patterns[i].rx, line, 0, NULL, 0) == 0)
			return 1;

	return 0;
}

static int grep_line(const char *fn, const char *line)
{
	int match;
	char numbuf[36];

	if (opt_match == MATCH_STRING)
		match = match_string(line);
	else
		match = match_regex(line);
	if (opt_match_invert)
		match = !match;

	n_lines++;
	if (match)
		n_matches++;

	switch (opt_output) {
	case OUT_NONE:
		return match ? STOP_LOOP : 0;
	case OUT_PATH:
		if (match) {
			printf("%s\n", fn);
			return STOP_LOOP;
		}
		/* fall through */
	case OUT_COUNT:
		return 0;
	case OUT_CONTENT:
		if (!match)
			return 0;
		break;
	}

	if (output_opts & OUT_LINENO)
		sprintf(numbuf, "%" PRIuMAX ":", n_lines);
	else
		numbuf[0] = 0;

	if (n_files > 1)
		printf("%s:%s%s\n", fn, numbuf, line);
	else
		printf("%s%s\n", numbuf, line);

	return 0;
}

static int grep_file(struct walker *w, const char *pr_fn, FILE *f)
{
	n_lines = 0;
	n_matches = 0;

	setlinebuf(f);

	while (1) {
		char *line;
		size_t len;
		int ret;

		line = fgets_unlocked(linebuf, sizeof(linebuf), f);
		if (!line) {
			if (ferror(f))
				w->exit_status = 2;
			break;
		}

		len = strlen(line);
		if (line[len - 1] == '\n')
			line[len - 1] = 0;

		ret = grep_line(pr_fn, line);
		if (ret & STOP_LOOP)
			break;
	}

	if (opt_output == OUT_COUNT)
		printf("%s%s%" PRIuMAX "\n",
		       n_files > 1 ? pr_fn : "",
		       n_files > 1 ? ":" : "",
		       n_matches);

	total_matches += n_matches;

	return 0;
}

static int grep_init(struct walker *w, int argc, char **argv)
{
	pu_init();

	/*
	 * support 'fgrep', 'egrep' program names
	 */
	struct pathelem *pe = path_split(argv[0]);
	if (!strcmp(pe->basen, "egrep"))
		opt_match = MATCH_ERE;
	else if (!strcmp(pe->basen, "fgrep"))
		opt_match = MATCH_STRING;
	path_free(pe);

	return 0;
}

static int grep_pre_walk(struct walker *w)
{
	if (!n_patterns) {
		char *s = slist_shift(&w->strlist);
		if (s)
			add_pattern(s);
	}

	if (!n_patterns) {
		fprintf(stderr, _("no patterns specified"));
		return 2;
	}

	if (opt_match != MATCH_STRING)
		compile_patterns();

	n_files = w->strlist.len;

	return 0;
}

static int grep_post_walk(struct walker *w)
{
	/* if -q, there were errors, AND there was a match, return 0 */
	if (total_matches && opt_output == OUT_NONE)
		return 0;	/* match */
	if (w->exit_status > 1)
		return w->exit_status; /* error */
	if (total_matches)
		return 0;	/* match */
	return 1;		/* no match */
}

int main (int argc, char *argv[])
{
	walker.argp			= &argp;
	walker.flags			= WF_NO_FILES_STDIN;
	walker.init			= grep_init;
	walker.pre_walk			= grep_pre_walk;
	walker.post_walk		= grep_post_walk;
	walker.cmdline_file		= grep_file;
	return walk(&walker, argc, argv);
}
